### PR TITLE
Add protobuf export of token information and symbol references

### PR DESCRIPTION
**Pull request by tvsonar**

### CODE REVIEW

File: src/SonarQube.SonarLint.Runner/CSharp/SymbolReferenceCollector.cs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                        }
+                    }
+                }
+            }
+        }
+
+        //From Roslyn (http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.CSharp.Workspaces/LanguageServices/CSharpSyntaxFactsService.cs,1453)
+        public SyntaxNode GetBindableParent(SyntaxToken token)
+        {
+            var node = token.Parent;
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/CSharp/SymbolReferenceCollector.cs]: MAJOR The 'this' parameter (or 'Me' in Visual Basic) of 'SymbolReferenceCollector.GetBindableParent(SyntaxToken)' is never used. Mark the member as static (or Shared in Visual Basic) or use 'this'/'Me' in the method body or at least one property accessor, if appropriate. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AMarkMembersAsStatic)

File: src/SonarQube.SonarLint.Runner/Program.cs [hunk 2]
```diff
@@ -77,6 +77,9 @@ public static int Main(string[] args)
                         var compilation = solution.Projects.First().GetCompilationAsync().Result;
                         var syntaxTree = compilation.SyntaxTrees.First();
 
+                        var x = new CSharp.SymbolReferenceCollector(solution.GetDocument(syntaxTree), solution.Workspace);
+                        var references = x.Generate().ToList();
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Program.cs]: MAJOR Remove this unused "references" local variable. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [Program.cs#L103](https://github.com/SonarSource-VisualStudio/sonaranalyzer-dotnet/blob/adab20439eaa285ef9ebac4574f27678737fbf1b/src/SonarQube.SonarLint.Runner/Program.cs#L103): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+
+    #region Descriptor
+    /// <summary>File descriptor for ColorizerInfo.proto</summary>
+    public static pbr::FileDescriptor Descriptor {
+      get { return descriptor; }
+    }
+    private static pbr::FileDescriptor descriptor;
+
+    static ColorizerInfoReflection() {
+      byte[] descriptorData = global::System.Convert.FromBase64String(
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR Initialize all static fields in 'ColorizerInfoReflection' when those fields are declared and remove the explicit static constructor. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AInitializeReferenceTypeStaticFieldsInline)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+            private int end_;
+            public int End {
+              get { return end_; }
+              set {
+                end_ = value;
+              }
+            }
+
+            public override bool Equals(object other) {
+              return Equals(other as Location);
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR In member 'ColorInfo.Types.ColorInfoInFile.Types.Location.Equals(object)', change parameter name 'other' to 'obj' in order to match the identifier as it has been declared in 'object.Equals(object)'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AParameterNamesShouldMatchBaseDeclaration)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+    public const int ColorInfosFieldNumber = 1;
+    private static readonly pb::FieldCodec<global::SonarAnalyzer.Protobuf.ColorInfo> _repeated_colorInfos_codec
+        = pb::FieldCodec.ForMessage(10, global::SonarAnalyzer.Protobuf.ColorInfo.Parser);
+    private readonly pbc::RepeatedField<global::SonarAnalyzer.Protobuf.ColorInfo> colorInfos_ = new pbc::RepeatedField<global::SonarAnalyzer.Protobuf.ColorInfo>();
+    public pbc::RepeatedField<global::SonarAnalyzer.Protobuf.ColorInfo> ColorInfos {
+      get { return colorInfos_; }
+    }
+
+    public override bool Equals(object other) {
+      return Equals(other as ColorInfoCollection);
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR In member 'ColorInfoCollection.Equals(object)', change parameter name 'other' to 'obj' in order to match the identifier as it has been declared in 'object.Equals(object)'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AParameterNamesShouldMatchBaseDeclaration)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+                size += 1 + pb::CodedOutputStream.ComputeInt32Size(Start);
+              }
+              if (End != 0) {
+                size += 1 + pb::CodedOutputStream.ComputeInt32Size(End);
+              }
+              return size;
+            }
+
+            public void MergeFrom(Location other) {
+              if (other == null) {
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR In member 'ColorInfo.Types.ColorInfoInFile.Types.Location.MergeFrom(ColorInfo.Types.ColorInfoInFile.Types.Location)', change parameter name 'other' to 'message' in order to match the identifier as it has been declared in 'IMessage<ColorInfo.Types.ColorInfoInFile.Types.Location>.MergeFrom(ColorInfo.Types.ColorInfoInFile.Types.Location)'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AParameterNamesShouldMatchBaseDeclaration)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+      int size = 0;
+      if (FilePath.Length != 0) {
+        size += 1 + pb::CodedOutputStream.ComputeStringSize(FilePath);
+      }
+      size += colorInfos_.CalculateSize(_repeated_colorInfos_codec);
+      return size;
+    }
+
+    public void MergeFrom(ColorInfo other) {
+      if (other == null) {
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR In member 'ColorInfo.MergeFrom(ColorInfo)', change parameter name 'other' to 'message' in order to match the identifier as it has been declared in 'IMessage<ColorInfo>.MergeFrom(ColorInfo)'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AParameterNamesShouldMatchBaseDeclaration)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+    }
+
+    public int CalculateSize() {
+      int size = 0;
+      size += colorInfos_.CalculateSize(_repeated_colorInfos_codec);
+      return size;
+    }
+
+    public void MergeFrom(ColorInfoCollection other) {
+      if (other == null) {
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR In member 'ColorInfoCollection.MergeFrom(ColorInfoCollection)', change parameter name 'other' to 'message' in order to match the identifier as it has been declared in 'IMessage<ColorInfoCollection>.MergeFrom(ColorInfoCollection)'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AParameterNamesShouldMatchBaseDeclaration)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+    public const int ColorInfosFieldNumber = 2;
+    private static readonly pb::FieldCodec<global::SonarAnalyzer.Protobuf.ColorInfo.Types.ColorInfoInFile> _repeated_colorInfos_codec
+        = pb::FieldCodec.ForMessage(18, global::SonarAnalyzer.Protobuf.ColorInfo.Types.ColorInfoInFile.Parser);
+    private readonly pbc::RepeatedField<global::SonarAnalyzer.Protobuf.ColorInfo.Types.ColorInfoInFile> colorInfos_ = new pbc::RepeatedField<global::SonarAnalyzer.Protobuf.ColorInfo.Types.ColorInfoInFile>();
+    public pbc::RepeatedField<global::SonarAnalyzer.Protobuf.ColorInfo.Types.ColorInfoInFile> ColorInfos {
+      get { return colorInfos_; }
+    }
+
+    public override bool Equals(object other) {
+      return Equals(other as ColorInfo);
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR In member 'ColorInfo.Equals(object)', change parameter name 'other' to 'obj' in order to match the identifier as it has been declared in 'object.Equals(object)'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AParameterNamesShouldMatchBaseDeclaration)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+        private global::SonarAnalyzer.Protobuf.ColorInfo.Types.ColorInfoInFile.Types.Location location_;
+        public global::SonarAnalyzer.Protobuf.ColorInfo.Types.ColorInfoInFile.Types.Location Location {
+          get { return location_; }
+          set {
+            location_ = value;
+          }
+        }
+
+        public override bool Equals(object other) {
+          return Equals(other as ColorInfoInFile);
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR In member 'ColorInfo.Types.ColorInfoInFile.Equals(object)', change parameter name 'other' to 'obj' in order to match the identifier as it has been declared in 'object.Equals(object)'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AParameterNamesShouldMatchBaseDeclaration)

File: src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+            size += 1 + pb::CodedOutputStream.ComputeStringSize(TokenType);
+          }
+          if (location_ != null) {
+            size += 1 + pb::CodedOutputStream.ComputeMessageSize(Location);
+          }
+          return size;
+        }
+
+        public void MergeFrom(ColorInfoInFile other) {
+          if (other == null) {
```
Comment (SonarTech) [inline src/SonarQube.SonarLint.Runner/Protobuf/ColorizerInfo.cs]: MAJOR In member 'ColorInfo.Types.ColorInfoInFile.MergeFrom(ColorInfo.Types.ColorInfoInFile)', change parameter name 'other' to 'message' in order to match the identifier as it has been declared in 'IMessage<ColorInfo.Types.ColorInfoInFile>.MergeFrom(ColorInfo.Types.ColorInfoInFile)'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AParameterNamesShouldMatchBaseDeclaration)

**Comment by SonarTech** [comment]
SonarQube analysis reported 17 issues
* MAJOR 16 major
* INFO 1 info

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:79F0B359-8275-4F3B-B919-82C493B74665: Do not nest type 'ColorInfo.Types'. Alternatively, change its accessibility so that it is not externally visible. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ANestedTypesShouldNotBeVisible)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:79F0B359-8275-4F3B-B919-82C493B74665: Do not nest type 'ColorInfo.Types.ColorInfoInFile'. Alternatively, change its accessibility so that it is not externally visible. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ANestedTypesShouldNotBeVisible)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:79F0B359-8275-4F3B-B919-82C493B74665: Do not nest type 'ColorInfo.Types.ColorInfoInFile.Types'. Alternatively, change its accessibility so that it is not externally visible. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ANestedTypesShouldNotBeVisible)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:79F0B359-8275-4F3B-B919-82C493B74665: Do not nest type 'ColorInfo.Types.ColorInfoInFile.Types.Location'. Alternatively, change its accessibility so that it is not externally visible. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ANestedTypesShouldNotBeVisible)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:79F0B359-8275-4F3B-B919-82C493B74665: Rename type name 'ColorInfoCollection' so that it does not end in 'Collection'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AIdentifiersShouldNotHaveIncorrectSuffix)
1. INFO [Program.cs#L103](https://github.com/SonarSource-VisualStudio/sonaranalyzer-dotnet/blob/b5820186bb442d969a2e416041e47eb4b0a2339c/src/SonarQube.SonarLint.Runner/Program.cs#L103): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/SonarQube.SonarLint.Runner/Protobuf/FileTokenInfo.proto [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+message TextRange {
+  int32 start_line = 1;
+  // End line (inclusive)
+  int32 end_line = 2;
+  int32 start_offset = 3;
+  int32 end_offset = 4;
+}
+
+enum TokenType {
+	UNKNOWN = 0;
```
Comment (tvsonar) [inline src/SonarQube.SonarLint.Runner/Protobuf/FileTokenInfo.proto]: Fix indenting.

File: src/SonarQube.SonarLint.Runner/TokenCollector/TokenHelper.cs [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    public static class TokenHelper
+    {
+        // From Roslyn (http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.CSharp.Workspaces/LanguageServices/CSharpSyntaxFactsService.cs,1453)
+        public static SyntaxNode GetBindableParent(this SyntaxToken token)
+        {
+            return token.Language == LanguageNames.CSharp
+                ? GetBindableParentCs(token)
+                : GetBindableParentVb(token);
+        }
+
```
Comment (tvsonar) [inline src/SonarQube.SonarLint.Runner/TokenCollector/TokenHelper.cs]: Remove duplicate empty line.

**Comment by tvsonar** [comment]
Superseded by https://github.com/SonarSource-VisualStudio/sonaranalyzer-dotnet/pull/1026

**Pull request closed without merging**